rclcpp/src/rclcpp/node_interfaces/node_graph.cpp

```diff
@@ -215,6 +256,66 @@ NodeGraph::get_client_names_and_types_by_node(
   return services_and_types;
 }
 
+std::map<std::string, std::vector<std::string>>
+NodeGraph::get_action_client_names_and_types_by_node(
+    rclcpp::Node node,
+    const std::string & node_name,
+    const std::string & namespace_) const
+{
+  rcl_names_and_types_t action_names_and_types = rcl_get_zero_initialized_names_and_types();
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcl_ret_t ret = rcl_action_get_client_names_and_types_by_node(
+    node_base_->get_rcl_node_handle(),
+    &allocator,
+    node_name.c_str(),
+    namespace_.c_str(),
+    &action_names_and_types);
+  if (ret != RCL_RET_OK) {
+    throw_from_rcl_error(ret, "failed to get action names and types by node");
+  }
+
+  std::map<std::string, std::vector<std::string>> actions_and_types;
+  for (size_t i = 0; i < action_names_and_types.names.size; ++i) {
+    std::string action_name = action_names_and_types.names.data[i];
+    for (size_t j = 0; j < action_names_and_types.types[i].size; ++j) {
+      actions_and_types[action_name].emplace_back(action_names_and_types.types[i].data[j]);
+    }
+  }
+
+  return actions_and_types;
+}
+
+std::map<std::string, std::vector<std::string>>
+NodeGraph::get_action_server_names_and_types_by_node(
+    rclcpp::Node node,
+    const std::string & node_name,
+    const std::string & namespace_) const
+{
+  rcl_names_and_types_t action_names_and_types = rcl_get_zero_initialized_names_and_types();
```

Comment:
i think this needs to be finalized with rcl_names_and_types_fini before exiting from this function scope to avoid memory leak.